### PR TITLE
Translations differ between platforms

### PR DESCRIPTION
**Pull request by japsmits**
# Description

Ensure all translatable strings are visible to the translator, regardless of system configuration used.

## Related issues

#3562

# Manual testing

Started DOSBox and tried the changed commands.

# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [ ] commented on the particularly hard-to-understand areas of my code.
- [x] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [x] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [ ] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [ ] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

### CODE REVIEW

File: contrib/resources/translations/en.lng [hunk 1]
```diff
@@ -544,9 +544,19 @@ Select the frame presentation mode:
...
+  opengl:    Uses bilinear interpolation.
+.
+:CONFIGITEM_OUTPUT_TEXTURE
+  texture:   Uses bilinear interpolation.
+.
+:CONFIGITEM_OUTPUT_OPENGLNB
+  openglnb:  Uses nearest-neighbour (no-bilinear).
+.
+:CONFIGITEM_OUTPUT_TEXTURENB
+  texturenb: Uses nearest-neighbour (no-bilinear).
```
Comment (johnnovak) [inline contrib/resources/translations/en.lng:559]: Minor: we put at least two spaces between the semicolon separating the value and its description everywhere in the config as it looks nicer.

Comment (johnnovak) [reply contrib/resources/translations/en.lng:559]: `Uses nearest-neighbour interpolation (no bilinear).`

File: contrib/resources/translations/en.lng [hunk 2]
```diff
@@ -864,7 +874,7 @@ Software scalers are deprecated in favour of hardware-accelerated options:
   - If you used an advanced scaler, consider one of the 'glshader'
     options instead.
 .
-:CONFIG_GLSHADER
+:CONFIGITEM_GLSHADER_GLSHADER
```
Comment (johnnovak) [inline contrib/resources/translations/en.lng:877]: This could have remained `CONFIG_GLSHADER`, no?

File: contrib/resources/translations/en.lng [hunk 3]
```diff
@@ -1084,17 +1094,17 @@ Sound is still fully emulated in silent mode, but DOSBox outputs silence.
...
 :CONFIG_PREBUFFER
 How many milliseconds of sound to render on top of the blocksize
-(20 by default). Larger values might help with sound stuttering but the sound
+(25 by default). Larger values might help with sound stuttering but the sound
 will also be more lagged.
 .
 :CONFIG_NEGOTIATE
 Let the system audio driver negotiate possibly better sample rate and blocksize
-settings (enabled by default).
+settings (disabled by default).
```
Comment (johnnovak) [inline contrib/resources/translations/en.lng:1107]: These settings vary per platform: https://github.com/dosbox-staging/dosbox-staging/blob/main/src/hardware/mixer.cpp#L2882-L2893

File: include/setup.h [hunk 4]
```diff
@@ -135,13 +135,15 @@ class Property {
 	virtual ~Property() = default;
 
 	void Set_values(const std::vector<std::string>& in);
+	void Set_enabled_options(const std::vector<std::string>& in);
```
Comment (johnnovak) [inline include/setup.h:138]: SetEnabledOptions

File: include/setup.h [hunk 5]
```diff
@@ -135,13 +135,15 @@ class Property {
...
 	void SetDeprecatedWithAlternateValue(const char* deprecated_value,
 	                                     const char* alternate_value);
 
 	void Set_help(const std::string& str);
+	void Set_option_help(const std::string& option, const std::string& in);
```
Comment (johnnovak) [inline include/setup.h:143]: SetOptionHelp (we use UpperCamelCase for methods)

File: src/debug/debug_gui.cpp [hunk 6]
```diff
@@ -278,14 +278,14 @@ void LOG_StartUp(void) {
 	
 	/* Register the log section */
 	Section_prop * sect=control->AddSection_prop("log",LOG_Init);
-	Prop_string* Pstring = sect->Add_string("logfile",Property::Changeable::Always,"");
-	Pstring->Set_help("File where the log messages will be saved to");
+	Prop_string* pstring = sect->Add_string("logfile",Property::Changeable::Always,"");
+	pstring->Set_help("File where the log messages will be saved to");
```
Comment (johnnovak) [inline src/debug/debug_gui.cpp:282]: Minor: might as well reformat this block

Comment (johnnovak) [reply src/debug/debug_gui.cpp:282]: Might as well change this terrible sentence to `Path of the log file` 😅

File: src/debug/debug_gui.cpp [hunk 7]
```diff
@@ -278,14 +278,14 @@ void LOG_StartUp(void) {
...
 	char buf[64];
 	for (Bitu i = LOG_ALL + 1;i < LOG_MAX;i++) {
 		safe_strcpy(buf, loggrp[i].front);
 		lowcase(buf);
-		Prop_bool* Pbool = sect->Add_bool(buf,Property::Changeable::Always,true);
-		Pbool->Set_help("Enable/disable logging of this type.");
+		Prop_bool* pbool = sect->Add_bool(buf,Property::Changeable::Always,true);
+		pbool->Set_help("Enable/disable logging of this type.");
```
Comment (johnnovak) [inline src/debug/debug_gui.cpp:288]: Minor: reformat

File: src/gui/sdlmain.cpp [hunk 8]
```diff
@@ -4323,21 +4320,26 @@ static void config_add_sdl()
...
 #if C_OPENGL
-	Pstring = sdl_sec->Add_string("output", always, "opengl");
-	Pstring->Set_help(
-	        "Video system to use for output ('opengl' by default).\n"
-	        "'texture' and 'opengl' use bilinear interpolation, 'texturenb' and\n"
-	        "'openglnb' use nearest-neighbour (no-bilinear). Some shaders require\n"
-	        "bilinear interpolation, making that the safest choice.");
-	Pstring->SetDeprecatedWithAlternateValue("openglpp", "opengl");
-	Pstring->SetDeprecatedWithAlternateValue("surface", "opengl");
+	const std::string def_output = "opengl";
```
Comment (johnnovak) [inline src/gui/sdlmain.cpp:4323]: Maybe spell it out: `default_output`

File: src/gui/sdlmain.cpp [hunk 9]
```diff
@@ -4323,21 +4320,26 @@ static void config_add_sdl()
...
-	Pstring->SetDeprecatedWithAlternateValue("texturepp", "texture");
-	Pstring->Set_values({
+	pstring = sdl_sec->Add_string("output", always, def_output.c_str());
+	pstring->Set_help(
+	        "Video system to use for output ('" + def_output + "' by default).\n"
+	        "Some shaders require bilinear interpolation, making that the safest choice.");
+	pstring->Set_option_help("opengl",    "  opengl:    Uses bilinear interpolation.");
+	pstring->Set_option_help("texture",   "  texture:   Uses bilinear interpolation.");
+	pstring->Set_option_help("openglnb",  "  openglnb:  Uses nearest-neighbour (no-bilinear).");
+	pstring->Set_option_help("texturenb", "  texturenb: Uses nearest-neighbour (no-bilinear).");
```
Comment (johnnovak) [inline src/gui/sdlmain.cpp:4334]: Minor: two spaces minimum between options and their descriptions.

File: contrib/resources/translations/en.lng [hunk 10]
```diff
@@ -2137,7 +2192,7 @@ Notes:
     for CD-based games that need a real DOS environment via bootable HDD image.
 
 Examples:
-  [color=light-green]imgmount[reset] [color=white]D[reset] [color=light-cyan]/Users/USERNAME/Games/doom.iso[reset] -t cdrom
+  [color=light-green]imgmount[reset] [color=white]D[reset] [color=light-cyan]C:\Games\doom.iso[reset] -t cdrom
```
[the remark this answers is not in the document]
Comment (japsmits) [reply contrib/resources/translations/en.lng:2195]: I agree this should be handled, but current code only supports settings for now. This instance specifies the help-text of a command, which takes a completely different code-path. I would suggest taking that up in another issue and PR.

@FeralChild64 and @johnnovak: is this only about paths being platform-dependent? If so, we could add a platform-independent example-path generator, and remove translations of paths from of the equation.

File: src/misc/setup.cpp [hunk 11]
```diff
@@ -267,25 +267,73 @@ bool Property::ValidateValue(const Value& in)
...
+void Property::Set_help(const std::string& in)
 {
-	std::string result = "CONFIG_" + propname;
-	upcase(result);
-	return MSG_Get(result.c_str());
+	MSG_Add(create_config_name(propname).c_str(), in.c_str());
 }
 
-const char* Property::GetHelpUtf8() const
+static std::string create_configitem_name(const std::string& propname,
```
Comment (johnnovak) [inline src/misc/setup.cpp:282]: Minor: I'd prefer `config_item` everywhere (two words)

File: src/dos/program_mount.cpp [hunk 12]
```diff
@@ -72,6 +72,20 @@ void MOUNT::ListMounts()
...
+{
+	MoreOutputStrings output(*this);
+	output.AddString(MSG_Get("PROGRAM_MOUNT_HELP_LONG"));
+#ifdef WIN32
+	output.AddString(MSG_Get("PROGRAM_MOUNT_HELP_LONG_WIN32"));
+#elif MACOSX
+	output.AddString(MSG_Get("PROGRAM_MOUNT_HELP_LONG_MACOSX"));
+#else
+	output.AddString(MSG_Get("PROGRAM_MOUNT_HELP_LONG_OTHER"));
+#endif
```
Comment (johnnovak) [inline src/dos/program_mount.cpp:85]: This does the job nicely 👍🏻 Inventing a generic mechanism for only these two occurrences would be hardly warranted.

File: contrib/resources/translations/en.lng [hunk 13]
```diff
@@ -542,11 +542,25 @@ Select the frame presentation mode:
...
+  opengl:     Uses bilinear interpolation.
+.
+:CONFIGITEM_OUTPUT_TEXTURE
+  texture:    Uses bilinear interpolation.
+.
+:CONFIGITEM_OUTPUT_OPENGLNB
+  openglnb:   Uses nearest-neighbour interpolation (no-bilinear).
+.
+:CONFIGITEM_OUTPUT_TEXTURENB
+  texturenb:  Uses nearest-neighbour interpolation (no-bilinear).
```
Comment (johnnovak) [inline contrib/resources/translations/en.lng:563]: Minor: just "no bilinear" (drop the `-` please).

Comment (japsmits) [reply contrib/resources/translations/en.lng]: Did a find/replace, which explains the commit to vga.h.

File: src/dosbox.cpp [hunk 14]
```diff
@@ -1240,17 +1240,19 @@ void DOSBOX_Init()
...
+#if C_IPX
+	pbool->SetEnabledOptions({"ipx"});
 #endif
 
-#if C_SLIRP
 	secprop = control->AddSection_prop("ethernet", &NE2K_Init, changeable_at_runtime);
 
 	pbool = secprop->Add_bool("ne2000", when_idle, true);
-	pbool->Set_help(
+	pbool->SetOptionHelp(
```
Comment (johnnovak) [inline src/dosbox.cpp:1262]: Wouldn't the `ethernet` section be present then even if we're compiling without SLIRP support? I think the old `#if` needs to stay in this case, wrapping the whole section definition.

File: src/hardware/mixer.cpp [hunk 15]
```diff
@@ -2906,28 +2906,25 @@ void init_mixer_dosbox_settings(Section_prop& sec_prop)
...
+	        "(%s by default). Larger values might help with sound stuttering but the sound\n"
+	        "will also be more lagged.");
 
 	bool_prop = sec_prop.Add_bool("negotiate", only_at_start, default_allow_negotiate);
-	bool_prop->Set_help(
-	        format_str("Let the system audio driver negotiate possibly better sample rate and blocksize\n"
-	                      "settings (%s by default).",
-	                      default_allow_negotiate ? "enabled" : "disabled"));
+	bool_prop->Set_help("Let the system audio driver negotiate possibly better sample rate and blocksize\n"
+	                   "settings (%s by default).");
```
Comment (johnnovak) [inline src/hardware/mixer.cpp:2927]: Minor: formatting

File: src/midi/midi.cpp [hunk 16]
```diff
@@ -725,93 +725,100 @@ void init_midi_dosbox_settings(Section_prop& secprop)
...
 	        "  - This option has no effect when using the built-in synthesizers\n"
-	        "    ('mididevice = fluidsynth' or 'mididevice = mt32').\n"
+	        "    ('mididevice = fluidsynth' or 'mididevice = mt32').");
+	str_prop->SetOptionHelp(
+	        "coreaudio",
+	        "  - When using 'coreaudio', you can specify a SoundFont here.");
+	str_prop->SetOptionHelp("alsa",
+	                            "  - When using ALSA, use the Linux command 'aconnect -l' to list all open\n"
+	                            "    MIDI ports and select one (e.g. 'midiconfig = 14:0' for sequencer\n"
+	                            "    client 14, port 0).");
```
Comment (johnnovak) [inline src/midi/midi.cpp:798]: Minor: indentation / formatting

File: contrib/resources/translations/en.lng [hunk 17]
```diff
@@ -1081,20 +1095,20 @@ Enable silent mode (disabled by default).
...
 .
 :CONFIG_BLOCKSIZE
-Mixer block size in sample frames (512 by default). Larger values might help
+Mixer block size in sample frames (%s by default). Larger values might help
 with sound stuttering but the sound will also be more lagged.
 .
 :CONFIG_PREBUFFER
 How many milliseconds of sound to render on top of the blocksize
-(20 by default). Larger values might help with sound stuttering but the sound
+(%s by default). Larger values might help with sound stuttering but the sound
```
Comment (johnnovak) [inline contrib/resources/translations/en.lng:1106]: Very nice!

File: contrib/resources/translations/en.lng [hunk 18]
```diff
@@ -1140,27 +1154,60 @@ Note: You can fine-tune each channel's chorus level using the MIXER.
...
   mt32:        The built-in Roland MT-32 synthesizer.
                See the [mt32] section for detailed configuration.
+.
+:CONFIGITEM_MIDIDEVICE_AUTO
   auto:        Either one of the built-in MIDI synthesisers (if `midiconfig` is
                set to 'fluidsynth' or 'mt32'), or a MIDI device external to
                DOSBox (any other 'midiconfig' value). This might be a software
                synthesizer or physical device. This is the default behaviour.
+.
+:CONFIGITEM_MIDIDEVICE_NONE
```
Comment (johnnovak) [inline contrib/resources/translations/en.lng:1188]: No more flip flopping in `en.lng` when we generate it on different OSes🎖️

File: contrib/resources/translations/en.lng [hunk 19]
```diff
@@ -2759,10 +2841,25 @@ Notes:
...
   [color=light-green]mount[reset] [color=white]C[reset] [color=light-cyan]~/dosgames[reset]
   [color=light-green]mount[reset] [color=white]D[reset] [color=light-cyan]"/Volumes/Game CD"[reset] -t cdrom -usecd 0
   [color=light-green]mount[reset] [color=white]C[reset] [color=light-cyan]my_savegame_files[reset] -t overlay
 
+.
+:PROGRAM_MOUNT_HELP_LONG_OTHER
+  [color=light-green]mount[reset] [color=white]C[reset] [color=light-cyan]~/dosgames[reset]
+  [color=light-green]mount[reset] [color=white]D[reset] [color=light-cyan]"/media/USERNAME/Game CD"[reset] -t cdrom -usecd 0
+  [color=light-green]mount[reset] [color=white]C[reset] [color=light-cyan]my_savegame_files[reset] -t overlay
+
```
Comment (johnnovak) [inline contrib/resources/translations/en.lng:2862]: Nice 👍🏻

File: src/misc/setup.cpp [hunk 20]
```diff
@@ -302,7 +302,16 @@ void Property::SetOptionHelp(const std::string& in)
...
 {
-	std::string result = MSG_Get(create_config_name(propname).c_str());
+	std::string result = {};
+	if (MSG_Exists(create_config_name(propname).c_str())) {
+		std::string help_text = MSG_Get(create_config_name(propname).c_str());
+		// Fill in the default value if the help text contains '%s'.
+		if (help_text.find("%s") != std::string::npos) {
+			help_text = format_str(help_text,
+			                       GetDefaultValue().ToString().c_str());
+		}
```
Comment (johnnovak) [inline src/misc/setup.cpp:312]: Nice, I like it! 🚀

File: src/misc/setup.cpp [hunk 21]
```diff
@@ -324,12 +333,25 @@ std::string Property::GetHelp() const
 			result.append("\n");
 		}
 	}
+	if (result.empty()) {
+		LOG_WARNING("CONFIG: No help available for '%s'.", propname.c_str());
+		return "No help available for '" + propname + "'\n";
```
Comment (johnnovak) [inline src/misc/setup.cpp:338]: 💯

**Review by johnnovak — APPROVED** [review]
Very nice enhancement @japsmits, and bonus points for the helpful warning messages and other various usability enhancements! 🎉 Clearly, you've put a lot of thought into this. This will surely make the translators' lives easier.

File: src/dosbox.cpp (hunk 14 above)
Comment (japsmits) [reply src/dosbox.cpp:1262]: Fixed it. 😄

Comment (johnnovak) [reply src/dosbox.cpp:1262]: Very nice 😄

**Comment by johnnovak** [comment]
Don't want this to be held up by the MSCV issues and my 0.81.1 build experimentation, so merging now 🚀

**Pull request merged**